matchpygen/bipartite.h

```diff
@@ -0,0 +1,242 @@
+#ifndef MATCHPY_BIPARTITE_H
+#define MATCHPY_BIPARTITE_H
+
+#include <deque>
+#include <map>
+#include <set>
+#include <symengine/basic.h>
+#include <symengine/pow.h>
+#include <symengine/add.h>
+
+#include "common.h"
+
+using namespace std;
+
+typedef tuple<int, int> TItem; // either TLeft or TRight, use <variant> instead?
+typedef TItem TRight;
+typedef TItem TLeft;
+
+
+const int LEFT = 0;
+const int RIGHT = 1;
+
+typedef map<tuple<int, int>, tuple<int, int> > Matching;
+
+//(Generic[TLeft, TRight, TEdgeValue], MutableMapping[Tuple[TLeft, TRight], TEdgeValue])
+/*
+ * A bipartite graph representation.
+ */
+template<typename TEdgeValue>
+class BipartiteGraph {
+public:
+
+	//Node = Tuple[int, Union[TLeft, TRight]]
+	// typedef tuple<int, variant<TLeft, TRight>> Node; ???
+	typedef tuple<int, TItem> Node;
```

Comment:
`TItem` is either `TLeft` or `TRight`. We could have `typedef variant<TLeft, TRight> TItem`, but this is probably only supported in C++17. As an alternative, two separate fields could be used. This latter alternative would probably some quantity of code to be reviewed.